### ISSUE TITLE
UMI Quality 

### ISSUE THREAD

**Issue by prot-devin** [fulcrumgenomics/fgbio#1008]
Is there a quality metric associated with the umi / barcode grouping? e.g. "UMI1" is identified to be a set of n reads using some strategy. In the simpliest case of identity for example, all have the same UMI read, but of varying quality. Is there a set of metrics to describe the confidence of the grouping to UMI1? This is a little complicated as it of course depends on not only the DNA sequence quality, but also the UMI length (and if we're really going to model error "correctly" the representation of bases or other UMIs and their edit distances).

What I'd ideally like is a likelihood ratio of the read being associated with a particular UMI grouping vs the next nearest group per-read. This can be calculated from the phred scores of the read and the nearest observed alternative UMI.

**Comment by cehtolonen** [comment]
Hi @prot-devin, sorry for the delay in responding.
Is this still an issue for you? Thanks!

**Comment by prot-devin** [comment]
Hi @cehtolonen yes, this is still a desirable feature for me.

**Comment by nh13** [comment]
@prot-devin I think this is a really interesting idea.  We take inspiration from [Smith et al](http://doi.org/10.1101/gr.209601.116) to do the grouping, which does not have a way of quantifying the "confidence we have in each grouping of a UMI".  I have ideas, and I'm sure others at Fulcrum would too, and we would welcome sponsorship from yourself or others who would like to see this implemented.

**Comment by yfarjoun** [comment]
Calculating the _likelihood_ that the read's umi comes from any given UMI-set is quite straightforward to calculate given the base qualities and the size of the UMI-set. turning that into a probability of "correctness" would require calculating that for "all" the possible UMI-sets and and figuring out what prior to use.... We would be ignoring "bulk edit" events such as index-hopping (for non-inline UMIs) and other disasters...

**Comment by yfarjoun** [comment]
This also made me think that a practical use-case for this value needs to be developed...for example: an option in Consensus calling to filter low probability reads.

**Comment by nh13** [comment]
I was thinking of a phred-scaled probability of reads in the set being from different molecules.  We would not only use the reads in the set, but other reads in sets with the same outer coordinates but similar (but not the same) UMIs.

**Comment by prot-devin** [comment]
Yes, exactly as @nh13  described. A per-read membership probability, and indeed @yfarjoun I'm imagining a consensus calling filter for members with lower probability. I would also like to just annotate the sequences with the confidence metric so I can weight model training downstream. This is admittedly a more niche use case.
